src/main/java/com/github/pgasync/impl/ConnectionValidator.java

```diff
@@ -26,26 +26,40 @@
 public class ConnectionValidator {
 
     final String validationQuery;
+    final boolean validateSocket;
 
-    public ConnectionValidator(String validationQuery) {
-        this.validationQuery = validationQuery;
+    public ConnectionValidator(String validationQuery, boolean validateSocket) {
+        // Trimmed as empty means no query for backwards compatibility
+        this.validationQuery = validationQuery == null || validationQuery.trim().isEmpty() ? null : validationQuery;
+        this.validateSocket = validateSocket;
     }
 
     public Observable<Connection> validate(Connection connection) {
-        return connection.queryRows(validationQuery)
-                .lift(subscriber -> new Subscriber<Row>() {
-                    @Override
-                    public void onError(Throwable e) {
-                        subscriber.onError(e);
-                    }
-                    @Override
-                    public void onCompleted() {
-                        subscriber.onNext(connection);
-                        subscriber.onCompleted();
-                    }
-                    @Override
-                    public void onNext(Row row) { }
-                });
+        Observable<Connection> ret = Observable.just(connection);
+        if (validationQuery != null) {
+            ret = ret.flatMap(conn -> connection.queryRows(validationQuery)
+                    .lift(subscriber -> new Subscriber<Row>() {
```

Comment:
how about instead of this code not do:

       return connection.queryRows(validationQuery)
                .map( __-> connection);
?

it will return connection, ignore the query result and in case of any errors propagate them forward.

Reply:
My goal here, in adding my socket validation, was to leave the existing logic as untouched as possible to avoid breaking any downstream code expectations. Therefore I simply refactored out what was there and added my check. Your comment applies to the original code as well and, to keep w/ proper refactoring principles, maybe should be addressed there (i.e. open a PR addressing those concerns w/ the original source).

Reply:
I agree with your assumptions. It looks tempting to just remove all this code here :)